lib/declarative_authorization/authorization.rb

```diff
@@ -340,11 +340,16 @@ def user_roles_privleges_from_options(privilege, options)
       raise AuthorizationUsageError, "No user object given for user id (#{user.try(:id)}) or " +
         "set through Authorization.current_user" unless user
 
-      roles = options[:user_roles] || flatten_roles(roles_for(user))
+      roles = options[:user_roles] || flattened_roles(user)
       privileges = flatten_privileges privileges, options[:context]
       [user, roles, privileges]
     end
 
+    def flattened_roles(user)
+      @flatten_roles ||= {}
+      @flatten_roles[user] ||= flatten_roles(roles_for(user))
```

Comment:
Is this instantiated per request?  If not couldn't this cause permissions to be mixed across vhosts?

Reply:
That's a good point. If not per-request, would need to namespace by account name.

Reply:
If so, we'd need to figure a way to make this work for everyone due to the gem being open-source.

Reply:
Perhaps the rails cache could be used here?  Any environment that uses hijacking should handle that automatically.

Reply:
That's a great idea

Reply:
Using Rails cache would work but I'm not clear on how we'd expire it. So maybe we could use RequestCache. However, both of these would be introducing knowledge of Rails deep inside the DeclarativeAuthorization which I'm not sure is that great as the code is Rails-agnostic currently (but maybe there is a way to make it work).

I was poking around and noticed that `flatten_roles(roles_for(user))` is only computed the `options[:user_roles]` are missing. So perhaps it would be possible to overwrite `def options_for_permit` in a base controller like so,
```
def options_for_permit(**)
  result = super
  user = result["user"]
  result[:user_roles] = RequestCache.fetch("user_roles_#{user.id}") do
    Authorization::Engine.roles_with_hierarchy_for(user)
  end
end
```

I haven't thought about it all the way, but maybe that's the way to fix it without touching ae_declarative_authorization???

Reply:
Actually, it is possible to use Rails.cache here as well ... with the cache expiration being user.updated_at and user_role.updated_at.

Reply:
In property app, the code is like:
```
def permitted_to?(privilege, context)
      Authorization::Engine.instance.permit?(privilege,
                                             { :user                => self,
                                               :object              => nil,
                                               :context             => context,
                                               :skip_attribute_test => true
                                             }
      )
    end
```
defined on `user` model. `permit?` method ultimately calls these methods in this gem. So for each user (no matter which vhost this user is on), it has its own `Authorization::Engine.instance`, so it might not be a problem?

Reply:
That instance is `@@instance`, so there is only one for the whole lifetime of the process, no?

But looking at this `permitted_to?`, we can also pass `:user_roles` to  `Authorization::Engine.instance.permit?` here.

Reply:
I think you are right, `@@instance` is class variable so it's the same one for the process.

And yes, we can pass in `user_roles` from property, which would help this line: https://github.com/appfolio/ae_declarative_authorization/blob/master/lib/declarative_authorization/authorization.rb#L343

Though the other usage here doesn't accept `user_roles`: https://github.com/appfolio/ae_declarative_authorization/blob/master/lib/declarative_authorization/authorization.rb#L278

Reply:
The naming is a bit messy here...the `options[:user_roles]` are actually `user.role_symbols`.

Reply:
Actually...in `@flatten_roles[user] ||= flatten_roles(roles_for(user))`, we use `user` object as key for the hash, so even for the multiple vhosts the same slice, as long as the `user` is different, I think it won't mix up permissions across vhosts?

Maybe we can use `user` and some timestamp as the hash key, so that we can "invalidate" the value when it's necessary. For example, the `updated_at` column for user (could be alternating user role for this user).

Reply:
ActiveRecord uses id to test for equality, so I think you'd end up mixing data for users with the same id in different vhosts.  You can try this in the rails console:
```ruby
m1 = SomeModel.new(id: 123)
m2 = SomeModel.new(id: 123)
h = {}
h[m1] = 'hello'
puts h[m2] # => 'hello'
```

Reply:
ohh I didn't know this...

What about using `#{vhost_name}_#{user_id}` as key?

Reply:
One other thing to keep in mind is that this would only end up being cached for a particular worker, of which there are many and they are created and destroyed all the time as the app's requirements change

Reply:
I played around with a couple of ideas here,
https://github.com/appfolio/apm_bundle/pull/29145

One (the second commit) is to cache flatten_roles based on roles (i.e. `@flatten_roles[roles] ||=`) which works quite well despite the hash key being a long array of symbols. The other (the third commit) one is to overwrite `user_roles_privleges_from_options` and ask the user for its flattened role_symbols (i.e. which can be stored in Rails.cache based on user + user_role cache key). 

And in either way, we can provide a much better implementation for `flatten_roles` which I demonstrated in the second commit with `compute_flatten_roles`. 

I also added caching to `flatten_privileges` as that ends up producing most objects once flatten_roles is no longer an issue.

@blowfishpro @jimherz @JingyuZ